ports/nrf/drivers/bluetooth/ble_drv.c

```diff
@@ -784,10 +824,21 @@ void ble_drv_connect(uint8_t * p_addr, uint8_t addr_type) {
     conn_params.conn_sup_timeout  = BLE_CONN_SUP_TIMEOUT;
 
     uint32_t err_code;
+#if (BLUETOOTH_SD == 132 && BLE_API_VERSION == 4)
+    uint8_t conn_tag = BLE_CONN_CFG_TAG_DEFAULT;
+    if ((err_code = sd_ble_gap_connect(&addr,
+                                       &scan_params,
+                                       &conn_params,
+                                       conn_tag)) != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError,
+                  "Can not connect. status: 0x" HEX2_FMT, (uint16_t)err_code));
```

Comment:
there is an upstream effort to replace OSError errors with numbers, see https://github.com/micropython/micropython/issues/3269

Reply:
This is a very good point! All nlr_raise raises the error code / return value from the SD for verbosity. I think it we should just raise a numeric error here, and rather wrap this text into BLE_DRIVER_LOG() instead, which is only used when debugging this module. Would you be fine that i raise a new issue to make change s on all nlr_raise's in this file, keeping this nlr_raise as is for now?

As a side note, does MP_EIO make sense for SD error codes? or should we even make a mapping between nrf_error.h and mperrno.h error codes?

Reply:
created #149.

Reply:
regarding separate PR: good idea, this isn't high priority

regarding error codes: it really depends on whether you can hit it under normal operation or not. Similar to the "errors vs exceptions" separation in modern programming languages.
My view: errors that are expected to occur and must be handled by the caller (like "connection lost") should raise a descriptive error (even if it's a number). Errors that are unexpected and are really a programming error in the API (like BLE_ERROR_INVALID_CONN_HANDLE or NRF_ERROR_INVALID_STATE) should be checked and thrown as a generic error.
For some background, see [this great post](https://dave.cheney.net/2012/01/18/why-go-gets-exceptions-right) (in a somewhat different context).